server/src/Icepeak/Server/WebsocketServer/MultiSubscription.hs

```diff
@@ -0,0 +1,389 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Icepeak.Server.WebsocketServer.MultiSubscription (handleClient) where
+
+
+import Control.Concurrent.MVar (MVar)
+import qualified Control.Concurrent.Async as Async
+import qualified Control.Concurrent.MVar  as MVar
+
+import qualified Control.Exception as Exception
+import qualified Control.Monad     as Monad
+
+import Data.Functor  ((<&>))
+import qualified Data.Maybe    as Maybe
+
+import Data.HashMap.Strict (HashMap)
+import qualified Data.HashMap.Strict as HashMap
+
+import Data.Text (Text)
+import qualified Data.Text          as Text
+import qualified Data.Text.Encoding as Text
+
+import Data.Aeson (Value, (.=))
+import qualified Data.Aeson as Aeson
+
+import Data.UUID (UUID)
+import qualified Data.Time.Clock.POSIX as Clock
+import qualified System.Random         as Random
+
+import qualified Web.JWT               as JWT
+import qualified Network.WebSockets    as WS
+
+
+import Icepeak.Server.WebsocketServer.Payload
+import Icepeak.Server.Core (Core)
+import Icepeak.Server.Store (Path)
+import Icepeak.Server.Config (Config)
+import qualified Icepeak.Server.Config                  as Config
+import qualified Icepeak.Server.Core                    as Core
+import qualified Icepeak.Server.Metrics                 as Metrics
+import qualified Icepeak.Server.Subscription            as Subscription
+import qualified Icepeak.Server.JwtAuth                 as JwtAuth
+import qualified Icepeak.Server.AccessControl           as Access
+
+-- * Client handling
+
+newUUID :: IO UUID
+newUUID = Random.randomIO
+
+-- ** Sending Response Payloads
+
+data Client = Client
+  { clientConn :: WS.Connection
+  , clientUuid :: UUID
+  , clientCore :: Core
+  , clientIsDirty :: MVar ()
+  , clientSubscriptions :: MVar (HashMap Path (MVar Value))
```

Comment:
Is the top-level MVar in `clientSubscriptions` necessary? It looks like there
will only ever be one thread interacting with it, the client thread. The
producer always writes into the inner `MVar`.

Reply:
There's at least 2 threads interacting with it, e.g:
- updateThread
- onMessage
They both use the keys of the HashMap to do their business: e.g
- updateThread: takeMVarUpdatedValues
- onMessage: subscribes